File: 1371-find-the-longest-substring-containing-vowels-in-even-counts/1371-find-the-longest-substring-containing-vowels-in-even-counts.php
class Solution {
  function findTheLongestSubstring($s) {
    $v = [ 'a' => 1, 'e' => 2, 'i' => 4, 'o' => 8, 'u' => 16 ];
    $m = [ 0 => -1 ];
    $mask = 0;
    $ans = 0;
    foreach (str_split($s) as $i => $ch) {
      if (array_key_exists($ch, $v)) $mask ^= $v[$ch];
      if (array_key_exists($mask, $m)) {
        if ($ans < $i - $m[$mask]) $ans = $i - $m[$mask];
      } else {
        $m[$mask] = $i;
      }
    }
    return $ans;
  }
}

File: 1371-find-the-longest-substring-containing-vowels-in-even-counts/1371-find-the-longest-substring-containing-vowels-in-even-counts.py
class Solution:
  def findTheLongestSubstring(self, s: str) -> int:
    v = { 'a': 1 << 0, 'e': 1 << 1, 'i': 1 << 2, 'o': 1 << 3, 'u': 1 << 4 }
    m = { 0: -1 }
    mask = 0
    ans = 0
    for i in range(len(s)):
      if s[i] in v:
        mask ^= v[s[i]]
      if mask in m:
        if ans < i - m[mask]:
          ans = i - m[mask]
      else:
        m[mask] = i
    return ans

        